### PR TITLE
Add delmic load function

### PR DESCRIPTION
**Pull request by noemiebonnet**
### Description of the change
- Addition of a delmic reader 

### Progress of the PR
- [x] basic parsing of the file
- [x] basic testing
- [x] add metadata handling
- [x] ... (additional functionalities)
- [x] add tests,
- [x] update docstring (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/rosettasciio/blob/main/upcoming_changes/README.rst)),
- [x] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)
- [x] ready for review.

### CODE REVIEW

**Comment by jlaehne** [comment]
Tests are failing, because `docstrings` has been made private `_docstrings` in #132. Could you rename that in line #4 of your `_api.py`

**Comment by noemiebonnet** [comment]
Hi Jonas,

Yes I'll do it later today, sorry about that.

Best,
Noémie

Le mar. 13 juin 2023 à 17:00, Jonas Lähnemann ***@***.***> a
écrit :

**Comment by jlaehne** [comment]
Thanks for putting this forward @noemiebonnet 

As far as I understand, it only reads spectral images and not other types of Delmic-HDF5 files (e.g. single spectra, intensity maps, angle-resolved data) so far?

Note that we changed the placement and handling of the test data files in #123 

Concerning the metadata, it would be great if the full metadata tree from the hdf5 file could be placed under `original_metadata`.

For the fields mentioned in https://docs.lumispy.org/en/latest/user_guide/metadata_structure.html it would be great if they can be mapped to the `metadata` tree at some point.

File: rsciio/delmic/_api.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        },
+        {
+            "name": "Wavelength",
+            "axis": np.array(Acquisition2_ImageData_DimensionScaleC),
+            "units": "nm",
+            "navigate": False,
+        },
+    ]
+
+    metadata = {"signal": {"signal_type": "", "quantity": "Intensity (counts)"}}
```
Comment (jlaehne) [inline rsciio/delmic/_api.py:61]: If LumiSpy is installed, `signal_type` should be set to `CL` to provide the CL specific functions.

**Comment by jlaehne** [comment]
I added a progress tracker to the initial comment. It helps show others how far this PR is, feel free to amend the list by editing the comment.

**Comment by jlaehne** [comment]
pre-commit.ci autofix

**Comment by jlaehne** [comment]
pre-commit.ci autofix

File: doc/user_guide/supported_formats/delmic.rst [hunk 2]
```diff
@@ -0,0 +1,19 @@
+.. _delmic-format:
+
+Delmic formats
+---------------------
```
Comment (jlaehne) [inline doc/user_guide/supported_formats/delmic.rst:4]:
```suggestion
Delmic
------
```

File: doc/user_guide/supported_formats/delmic.rst [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    ``reader`` argument to define the correct file plugin as the ``.h5``
+    extension is not unique to this reader:
+
+    .. code-block:: python
+
+        >>> import hyperspy.api as hs
+        >>> hs.load("filename.h5", reader="Delmic")
+
+API functions
+"""""""""""""
```
Comment (jlaehne) [inline doc/user_guide/supported_formats/delmic.rst:19]:
```suggestion
"""""""""""""

.. automodule:: rsciio.delmic
   :members:
```

File: upcoming_changes/139.new.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+new: Support for Delmic HDF5 format.
```
Comment (jlaehne) [inline upcoming_changes/139.new.rst:1]:
```suggestion
Add basic support for the Delmic HDF5 format.
```

**Comment by noemiebonnet** [comment]
This PR contains only very simple and basic functionalities, more functionalities will be supported in the next pull request. Could you please merge it? The new PR is almost ready (end of this week)

File: rsciio/delmic/_api.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
+import h5py
+import numpy as np
+
+from rsciio._docstrings import FILENAME_DOC, RETURNS_DOC
+
+
+def file_reader(filename, *args, **kwds):
+    """Read a Delmic hdf5 hyperspectral image.
```
Comment (jlaehne) [inline rsciio/delmic/_api.py:8]:
```suggestion
    """
    Read a Delmic hdf5 hyperspectral image.

```

File: rsciio/delmic/_api.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
+import h5py
+import numpy as np
+
+from rsciio._docstrings import FILENAME_DOC, RETURNS_DOC
+
+
+def file_reader(filename, *args, **kwds):
```
Comment (jlaehne) [inline rsciio/delmic/_api.py:7]:
```suggestion
def file_reader(filename):

```

At the moment `*args` and `**kwds` are not used and just lead to a warning in building the documentation. If added again in the future, they should be documented in the docstring.

File: rsciio/delmic/_api.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
+import h5py
+import numpy as np
+
+from rsciio._docstrings import FILENAME_DOC, RETURNS_DOC
+
+
+def file_reader(filename):
```
Comment (jlaehne) [inline rsciio/delmic/_api.py:7]:
```suggestion
def file_reader(filename, lazy=False):

```

File: rsciio/delmic/_api.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
+import h5py
+import numpy as np
+
+from rsciio._docstrings import FILENAME_DOC, RETURNS_DOC
```
Comment (jlaehne) [inline rsciio/delmic/_api.py:4]:
```suggestion
from rsciio._docstrings import FILENAME_DOC, LAZY_DOC, RETURNS_DOC

```

File: rsciio/delmic/_api.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+def file_reader(filename, lazy=False):
+
+
+    """
+    Read a Delmic hdf5 hyperspectral image.
+
+
+    Parameters
+    ----------
+    %s
```
Comment (jlaehne) [inline rsciio/delmic/_api.py:16]:
```suggestion
    %s
    %s
```

File: rsciio/delmic/_api.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+        "metadata": metadata,
+        "original_metadata": original_metadata,
+    }
+
+    return [
+        spim,
+    ]
+
+
+file_reader.__doc__ %= (FILENAME_DOC, RETURNS_DOC)
```
Comment (jlaehne) [inline rsciio/delmic/_api.py:81]:
```suggestion
file_reader.__doc__ %= (FILENAME_DOC, LAZY_DOC, RETURNS_DOC)

```

**Comment by jlaehne** [comment]
pre-commit.ci autofix

**Pull request merged**